### PR TITLE
handle field / method name collision

### LINKED ISSUE #20: Field and Method that have same names are collinding.
I'm unable to access a static public method that have the same name as a static public field.

We might need some prefix to prevent fields and method from colliding.

### PR DESCRIPTION
**Pull request by olifozzy**
Allow interception when a field and a method have the same name.

note : I've added some unit tests that I wasn't able to run (no build env here)

Should fix #20 :)

### CODE REVIEW

**Comment by olifozzy** [comment]
Should fix #20 :)

File: lib/class-factory.js [hunk 1]
```diff
@@ -669,9 +669,13 @@ function ClassFactory (vm) {
           try {
             const fieldName = invokeObjectMethodNoArgs(env.handle, field, fieldGetName);
             try {
-              const fieldjsName = env.stringFromJni(fieldName);
+              var fieldjsName = env.stringFromJni(fieldName);
```
Comment (oleavr) [inline lib/class-factory.js]: Use `let` here to allow re-assigning – `var` is a legacy keyword from ES5.

File: lib/class-factory.js [hunk 2]
```diff
@@ -669,9 +669,13 @@ function ClassFactory (vm) {
...
               const fieldHandle = env.newGlobalRef(field);
               fieldHandles.push(fieldHandle);
+              // If we have a collided method, suffix the fieldName
+              if(jsMethods.hasOwnProperty(fieldjsName)){
```
Comment (oleavr) [inline lib/class-factory.js]: - Comment can be removed as it's already clear what the code is doing.
- Missing space after `if`.
- Theoretically there might be a method prefixed with `_` having the same name, so using `while` here would be better.

File: test/re/frida/MethodTest.java [hunk 3]
```diff
@@ -118,3 +174,17 @@ public int returnZero() {
...
+    
+    int Particle2 = 2;
+
+    int Particle() {
+        return 3;
+    }
+
+    static int Particle2(){
+        return 4;
+    }
```
Comment (oleavr) [inline test/re/frida/MethodTest.java]: Field- and method-names should follow the coding style, i.e. `camelCase`, not `CamelCase`.

**Comment by olifozzy** [comment]
Ok, I have fixed according to your feedback.

File: lib/class-factory.js [hunk 4]
```diff
@@ -669,9 +669,12 @@ function ClassFactory (vm) {
           try {
             const fieldName = invokeObjectMethodNoArgs(env.handle, field, fieldGetName);
             try {
-              const fieldjsName = env.stringFromJni(fieldName);
+              let fieldjsName = env.stringFromJni(fieldName);
               const fieldHandle = env.newGlobalRef(field);
               fieldHandles.push(fieldHandle);
+              while (jsMethods.hasOwnProperty(fieldjsName)){
```
Comment (oleavr) [inline lib/class-factory.js]: - Let's move these 3 lines to right after the `let`, as they belong together. Then put a blank line between that and the `fieldHandle` stuff (those two lines), and then one blank line before assigning to `jsFields`. Just to make the code a bit more readable.
- Missing a space before `{`.

File: test/re/frida/MethodTest.java [hunk 5]
```diff
@@ -118,3 +174,17 @@ public int returnZero() {
...
+class Collider {
+    static int particle = 1;
+    
+    int particle2 = 2;
+
+    int particle() {
+        return 3;
+    }
+
+    static int particle2(){
```
Comment (oleavr) [inline test/re/frida/MethodTest.java]: Missing space before `{`.

File: test/re/frida/MethodTest.java [hunk 6]
```diff
@@ -118,3 +174,17 @@ public int returnZero() {
         return 0;
     }
 }
+
+class Collider {
+    static int particle = 1;
+
```
Comment (oleavr) [inline test/re/frida/MethodTest.java]: Trailing whitespace.

File: test/re/frida/MethodTest.java [hunk 7]
```diff
@@ -65,6 +65,62 @@ public void genericsCanBeUsed() {
         assertEquals("Badger", script.getNextMessage());
     }
 
+
```
Comment (oleavr) [inline test/re/frida/MethodTest.java]: There's an extra blank line here.

**Comment by olifozzy** [comment]
Damn I miss stylecop ^^  Should be good now :)

File: lib/class-factory.js [hunk 8]
```diff
@@ -670,11 +670,13 @@ function ClassFactory (vm) {
...
-              const fieldHandle = env.newGlobalRef(field);
-              fieldHandles.push(fieldHandle);
-              while (jsMethods.hasOwnProperty(fieldjsName)){
+              while (jsMethods.hasOwnProperty(fieldjsName)) {
                 fieldjsName = "_" + fieldjsName;
               }
+
+              const fieldHandle = env.newGlobalRef(field);
+              fieldHandles.push(fieldHandle);
+
```
Comment (oleavr) [inline lib/class-factory.js]: Trailing whitespace.

**Comment by olifozzy** [comment]
Don't see any trailing whitespace :'(  Do you mean the extra line ?
trailing

**Comment by oleavr** [comment]
The last blank line isn't blank, it's got spaces before the CRLF.

**Pull request merged**